src/inference.jl

```diff
@@ -58,128 +58,51 @@ function Base.show(io::IO, mime::MIME"text/plain", model::LanguageModel)
 end
 
 struct KVCache
-    key_cache::Array{Float32,3}   # (head_size, n_heads, seq_len)
-    value_cache::Array{Float32,3} # (seq_len, head_size, n_heads)
+    key_cache::AbstractArray{Float32, 3}   # (head_size, n_heads, seq_len), {Float32,3}
+    value_cache::AbstractArray{Float32, 3} # (seq_len, head_size, n_heads), {Float32,3}
```

Comment:
Same parametric type thing here, though I think you might be able to do

```julia
struct KVCache{A<:AbstractArray{Float32, 3}}
    key_cache::A   # (head_size, n_heads, seq_len), {Float32,3}
    value_cache:: # (seq_len, head_size, n_heads), {Float32,3}
```

Reply:
I would think in this case Array would be Array{Float32, 3} ... also I wonder if this even compiles? :O We can use AbstractType with {} syntax? :astonished:

Reply:
Yes, I believe so. Abstract types can be parametric as well, like this one:

```julia
https://github.com/JuliaGraphs/Graphs.jl/blob/f97b4afd17e1cd938e53282bcb76c1b8e2ea127c/src/interface.jl#L34-L38
```

Reply:
Ah, I know what I confused this with. ParametricTypes couldn't be used with {} symbols.